librz/bin/format/dex/dex.c

```diff
@@ -1,239 +1,1855 @@
+
+#define dex_fail_if_eof(b, o, s, g) \
+	do { \
+		(o) = rz_buf_tell((b)); \
+		if ((o) >= (s)) { \
+			goto g; \
+		} \
+	} while (0)
+
+#define dex_fail_if_bad_ids(name, z, s, g) \
+	do { \
+		ut64 end = name##_offset; \
+		end += name##_size * (z); \
+		if (end >= (s)) { \
+			goto g; \
+		} \
+	} while (0)
+
+#define CLASS_ACCESS_FLAGS_SIZE 18
+static const DexAccessFlagsReadable access_flags_list[CLASS_ACCESS_FLAGS_SIZE] = {
+	{ ACCESS_FLAG_PUBLIC /*               */, "public" },
+	{ ACCESS_FLAG_PRIVATE /*              */, "private" },
+	{ ACCESS_FLAG_PROTECTED /*            */, "protected" },
+	{ ACCESS_FLAG_STATIC /*               */, "static" },
+	{ ACCESS_FLAG_FINAL /*                */, "final" },
+	{ ACCESS_FLAG_SYNCHRONIZED /*         */, "synchronized" },
+	{ ACCESS_FLAG_BRIDGE /*               */, "bridge" },
+	{ ACCESS_FLAG_VARARGS /*              */, "varargs" },
+	{ ACCESS_FLAG_NATIVE /*               */, "native" },
+	{ ACCESS_FLAG_INTERFACE /*            */, "interface" },
+	{ ACCESS_FLAG_ABSTRACT /*             */, "abstract" },
+	{ ACCESS_FLAG_STRICT /*               */, "strict" },
+	{ ACCESS_FLAG_SYNTHETIC /*            */, "synthetic" },
+	{ ACCESS_FLAG_ANNOTATION /*           */, "annotation" },
+	{ ACCESS_FLAG_ENUM /*                 */, "enum" },
+	{ ACCESS_FLAG_MODULE /*               */, "module" },
+	{ ACCESS_FLAG_CONSTRUCTOR /*          */, "constructor" },
+	{ ACCESS_FLAG_DECLARED_SYNCHRONIZED /**/, "synchronized" },
+};
 
-char *rz_bin_dex_get_version(RzBinDexObj *bin) {
-	rz_return_val_if_fail(bin, NULL);
-	char *version = calloc(1, 8);
-	if (version) {
-		rz_buf_read_at(bin->b, 4, (ut8 *)version, 3);
-		return version;
+static void dex_string_free(DexString *string) {
+	if (!string) {
+		return;
 	}
-	return NULL;
+	free(string->data);
+	free(string);
+}
+
+static DexString *dex_string_new(RzBuffer *buf, ut64 offset, st64 *pread) {
+	ut64 size = 0;
+	char *data = NULL;
+	st64 read;
+	DexString *string = NULL;
+
+	read = rz_buf_uleb128(buf, &size);
+	data = malloc(size + 1);
+	if (!data || rz_buf_read(buf, (ut8 *)data, size) != size) {
+		rz_warn_if_reached();
+		free(data);
+		return NULL;
+	}
+	data[size] = 0;
+
+	string = RZ_NEW0(DexString);
+	if (!string) {
+		rz_warn_if_reached();
+		free(data);
+		return NULL;
+	}
+
+	*pread = read;
+	string->size = size;
+	string->offset = offset;
+	string->data = data;
+	return string;
 }
 
-void rz_bin_dex_free(RzBinDexObj *dex) {
-	struct dex_header_t *dexhdr = &dex->header;
-	if (dex->cal_strings) {
-		size_t i;
-		for (i = 0; i < dexhdr->strings_size; i++) {
-			free(dex->cal_strings[i]);
+static void dex_proto_id_free(DexProtoId *string) {
+	if (!string) {
+		return;
+	}
+	free(string->type_list);
+	free(string);
+}
+
+static DexProtoId *dex_proto_id_new(RzBuffer *buf, ut64 offset) {
+	DexProtoId *proto_id = RZ_NEW0(DexProtoId);
+	if (!proto_id) {
+		return NULL;
+	}
+
+	ut32 parameters_offset = 0;
+
+	proto_id->shorty_idx = /*       */ rz_buf_read_le32(buf);
```

Comment:
not really, but makes the code more readable to me. and since it is something not exposed to the user (the one using the C api) then i do not think there is any reason to remove those.